### PR TITLE
Save state of DAO deployment.

### LINKED ISSUE #212: Pickup Where You Left Off: Maintain Internal Deployment State
In order to facilitate these two features:
- Pick-up deployment where you left off
- Recover from failed transactions

The migration script could maintain an internal state (think state-machine) of the DAO's setup process. This way if something goes wrong part way through, the current state can be serialized, and picked up once the problem is fixed.

This can be used for solving:
- dao-params.json issues that result in a crash
- Javascript bugs
- Failed transactions

### PR DESCRIPTION
**Pull request by ben-kaufman**
Fix #212

### CODE REVIEW

File: migrate-dao.js [hunk 1]
```diff
@@ -1,7 +1,22 @@
...
+
+function writeState (state) {
+  fs.writeFileSync(stateFile, JSON.stringify(state, undefined, 2), 'utf-8')
+}
 
 async function migrateDAO ({ web3, spinner, confirm, opts, migrationParams, logTx, previousMigration, customabislocation }) {
+  let deploymentState
+
+  if (fs.existsSync(stateFile)) {
+    deploymentState = JSON.parse(fs.readFileSync(stateFile))
```
Comment (orenyodfat) [inline migrate-dao.js]: I guess that this file need to be deleted after successful deployment .
also , there should be a special cmd something like `migrate -restart` which ignore or delete this file

Comment (ben-kaufman) [reply migrate-dao.js]: Yes it is deleted at the end. I will add a `-restart` flag

Comment (ben-kaufman) [reply migrate-dao.js]: Added

**Comment by ben-kaufman** [comment]
Hi, as for the use of the file system I'll add the callback as you asked.
I guess using TypeScript would've been a better choice in the beginning, but moving to it now is a lot of extra work which I currently don't have the time for. As we are currently working on refactoring Arc, and will probably therefore need to make many changes to the migration, that could might be a good time for such refactoring, but meanwhile I'll stick to JS...

I can maybe add some documentation for the state "schema" (though it's not even a schema, or at least wasn't intended to be), but I'm not sure what you mean as to "enforcing" it. It was meant to be written by the script itself, changed as the script changes and used "internally" within the script.
Also because the state of the script varies greatly according to the actual parameters, I'm not sure how I'd introduce this enforcement concept, and also the "steps" (not completely sure I understand what do you mean with that). 

I agree we could structure the script better now that we know the full requirements and capacities it needs to have, but this means rewriting it from scratch. Maybe we will do that when adapting for the new Arc, but right now that seems too time consuming.

**Comment by dOrgJelli** [comment]
By "enforced" I just meant type-checked.

By the steps I meant
```typescript
type DeploymentStep = (state: DeploymentState) => boolean
type DeploymentSteps = DeploymentStep[]
```

Within each step function, it would:
1. verify init state that was passed in
2. execute the deployment logic it's responsible for (ex: setting VM params)
3. modify the state object
4. return true if successful

The executor of these steps would do something like this:
```typescript
// psuedo code below
let deploymentState = ...;

steps.forEach((step, index) => {
  const newState = Object.clone(deploymentState)
  if (step(newState)) {
    deploymentState = newState;
    saveCurrentState(deploymentState, index)
  }
})
```

**Comment by ben-kaufman** [comment]
@dOrgJelli I separated the state storage from the script so you should now be able to use other methods for that.

I think I understand your suggestion, so as I said it's a good idea, just not for now as it isn't compatible with the present script.

**Pull request merged**